src/apps/vpn/controller.cpp

```diff
@@ -218,19 +219,25 @@ bool Controller::activate(const ServerData& serverData) {
 
   clearRetryCounter();
 
-  activateInternal();
+  activateInternal(DoNotForceDNSPort, serverSelectionPolicy);
   return true;
 }
 
-void Controller::activateInternal(bool forcePort53) {
+void Controller::activateInternal(DNSPortPolicy dnsPort,
+                                  ServerSelectionPolicy serverSelectionPolicy) {
   logger.debug() << "Activation internal";
   Q_ASSERT(m_impl);
 
   clearConnectedTime();
   m_handshakeTimer.stop();
   m_activationQueue.clear();
 
-  Server exitServer = Server::weightChooser(m_serverData.exitServers());
+  QList<Server> exitServers = m_serverData.exitServers();
+  Server exitServer = serverSelectionPolicy == DoNotRandomizeServerSelection &&
+                              !m_serverData.exitServerPublicKey().isEmpty()
+                          ? Server::publicKeyChooser(
```

Comment:
This could also just be done with `ServerCountryModel::server(const QString& pubkey)` which will do an O(1) hash lookup instead of needing to iterate through the server list.